components/scream/src/share/io/scorpio_output.hpp

```diff
@@ -0,0 +1,485 @@
+  // The other kind of output is the restart history output.  This allows a restart run to also have
+  // a consistent set if history outputs.
+  // 1. A restart history is not necessary for,
+  //   a. Instantaneous output streams.
+  //   b. When the history has also be written this step.  In other words, when the average counter is 0.
+  // Final point, typical output, a restart file and a restart history file can be distinguished by the
+  // suffix, (.nc) is typical, (.r.nc) is a restart and (.rhist.nc) is a restart history file.
+
+  // Check to see if output is expected and what kind.
+  bool is_typical = (m_status["Avg Count"] == m_out_frequency);
+  bool is_rhist   = (m_status["Avg Count"] == m_restart_hist_n) and !is_typical;
+  bool is_write = is_typical or is_rhist; // General flag for if output is written
+  // Preamble to writing output this step
+  std::string filename = m_casename+"."+m_avg_type+"."+m_out_units+"_x"+std::to_string(m_out_frequency)+"."+time_str;
+  if (m_is_restart) 
+  { 
+    filename+=".r";
+    std::ofstream rpointer;
+    rpointer.open("rpointer.atm",std::ofstream::out | std::ofstream::trunc);  // Open rpointer file and clear contents
+    rpointer << filename + ".nc" << std::endl;
+  }
+  if (is_rhist)
+  { 
+    filename+=".rhist"; 
+    std::ofstream rpointer;
+    rpointer.open("rpointer.atm",std::ofstream::app);  // Open rpointer file and append the restart hist file information
+    rpointer << filename + ".nc" << std::endl;
+    m_is_restart_hist = true;
+  }
+  filename += ".nc";
+  //
+  if (is_write) 
+  {
+    // If we closed the file in the last write because we reached max steps, or this is a restart history file,
+    // we need to create a new file for writing.
+    new_file(filename);
+    if( !m_is_init and is_typical ) { m_is_init=true; }
+
+    pio_update_time(filename,time); // Universal scorpio command to set the timelevel for this snap.
+    if (is_typical) { m_status["Snaps"] += 1; }  // Update the snap tally, used to determine if a new file is needed and only needed for typical output.
+  }
+
+  // Take care of updating and possibly writing fields.
+  for (auto const& f_map : m_fields)
+  {
+    // Get all the info for this field.
+    auto name   = f_map.first;
+    auto g_view = field_repo.get_field(f_map.second).get_view();
+    auto l_view = m_view.at(name); //TODO: may want to fix this since l_view may not be actually updated.  Use g_view for Instant?
+    Int  f_len  = f_map.second.get_layout().size();
+    // The next two operations do not need to happen if the frequency of output is instantaneous.
+    if (m_avg_type != "Instant")
+    {
+      // Update local view given the averaging type.  TODO make this a switch statement?
+      if (m_avg_type == "Average")
+      {
+        for (int ii=0; ii<f_len; ++ii) {l_view(ii) = (l_view(ii)*(m_status["Avg Count"]-1) + g_view(ii))/(m_status["Avg Count"]);}
```

Comment:
So if I understand correctly, `l_view` in this context is on Host and `g_view` is on Device, right?  Since we are keeping a persistent copy of `l_view` since it is updated from step to step the correct approach is to keep `l_view` on host and copy `g_view` to host for this operation.  Right?

Reply:
I believe this has been fixed now, see about 10 lines up.